plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -449,13 +500,30 @@ extendEnv :: Name -> CekValue uni fun -> CekValEnv uni fun -> CekValEnv uni fun
 extendEnv = insertByName
 
 -- | Look up a variable name in the environment.
-lookupVarName :: forall uni fun s . (PrettyUni uni fun) => Name -> CekValEnv uni fun -> CekM s (CekValue uni fun)
+lookupVarName :: forall uni fun s . (PrettyUni uni fun) => Name -> CekValEnv uni fun -> CekM uni fun s (CekValue uni fun)
 lookupVarName varName varEnv = do
     case lookupName varName varEnv of
-        Nothing  -> throwingWithCauseExc @(CekEvaluationException uni fun) _MachineError OpenTermEvaluatedMachineError $ Just var where
+        Nothing  -> throwingWithCause _MachineError OpenTermEvaluatedMachineError $ Just var where
             var = Var () varName
         Just val -> pure val
 
+-- | Take pieces of a possibly partial builtin application and either create a 'CekValue' using
+-- 'makeKnown' or a partial builtin application depending on whether the built-in function is
+-- fully saturated or not.
+evalBuiltinApp
```

Comment:
Maybe move this to after `applyEvaluate`?  `returnCek` calls `applyEvaluate` which is below it, but then that jumps up to here, before the entry point to the machine.  

Also, `applyEvaluate` and `forceEvaluate` are defined locally inside `enterCek`, but this is outside. Does that matter at all?  Are you still allowed `INLINE` in local definitions?

Reply:
> Also, `applyEvaluate` and `forceEvaluate` are defined locally inside `enterCek`, but this is outside. 

That's because there's this massive mutual recursion there, while `evalBuiltinApp` is not recursive at all and just constructs a value out of parts of a builtin application. It's a basic auxiliary function and I don't think it should be moved to the giant mutual block. That function is analogous to `emitCek` and `dischargeCekValue`, which we also call from the mutual block withoutt storing them there. 

> Are you still allowed INLINE in local definitions?

Yes.